### PR TITLE
Issue 44 wider repeat

### PR DESCRIPTION
**Pull request by obregos**
- add widerRepeated function
- delete dropRepeatingNA function

### CODE REVIEW

File: R/widerRepeated.R [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#'        \code{\link{exportRecordsTyped}}
+#' @param rcon A REDCap connection object as created by \code{unlockREDCap}.
+#' @param idvar \code{character}, Names of one or more variables in long format that identify multiple records belonging to the same group. 
+#'        These variables may also be present in wide format. (e.g., 'record_id')
+
+
+widerRepeated <- function(rcon, idvar){
+  # Records <- reshape(Records, idvar = idvar, timevar = timevar, direction = 'wide')
+  # export data dictionary
+  meta <- exportMetaData(rcon)
```
Comment (nutterb) [inline R/widerRepeated.R:16]: meta <- rcon$metadata()

File: R/widerRepeated.R [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  
+  # export repeating
+  repeating <- exportRepeatingInstrumentsEvents(rcon)
```
Comment (nutterb) [inline R/widerRepeated.R:19]: repeating <- rcon$repeatInstrumentEvent()

File: R/widerRepeated.R [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  
+  # export events
+  events <- exportEvents(rcon)
```
Comment (nutterb) [inline R/widerRepeated.R:22]: events <- rcon$events()

File: R/widerRepeated.R [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  
+  # export mappings
+  mappings <- exportMappings(rcon)
```
Comment (nutterb) [inline R/widerRepeated.R:25]: mappings <- rcon$mapping()

File: R/widerRepeated.R [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  dd <- data.frame(lapply(meta, as.character), stringsAsFactors = FALSE)
+  
+  # get correct event mappings if redcap_event_name is present or not
+  if ("redcap_event_name" %in% names(dat)) {
+    #get the event-form mapping
+    event.map <- mappings
+  } else {
+    event.map = data.frame(
+      unique_event_name = "",
+      form = sort(unique(dd$form_name)))
```
Comment (nutterb) [inline R/widerRepeated.R:46]: form = rcon$instruments()$instrument_name

File: R/widerRepeated.R [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  
+  id.fields = names(df.all_ids)
+  
+  map <- event.map
+  data.frame(map)
+  map[is.na(map)] <- ""
+  
+  for (i in sort(unique(dd$form_name))) {
+    # print(i)
+    vars.tmp <- names(dat)[sub("___.*", "", names(dat)) %in% dd$field_name[dd$form_name == i]]
```
Comment (nutterb) [inline R/widerRepeated.R:68]: This is _really_ obscure...using the following would be more consistent with how we handle checkbox field names elsewhere in the package

`sub(REGEX_CHECKBOX_FIELD_NAME, "\\1", names(dat))`

**Comment by nutterb** [comment]
I might be using this wrong, or misunderstanding the description.  When I run the following I was expecting to get one row for `id.tmp = 1` (with the two events spread into wide format).  

Everything leading up to the `for` loop looks great. I'm just unsure if this is a case of me misunderstanding of `reshape` being a pain in the butt.

```
> widerRepeated(rcon, "record_id")$repeating_instrument[1:2, ]
                    id.tmp redcap_event_name redcap_repeat_instance
1.repeat_question_1      1     event_1_arm_1                      1
2.repeat_question_1      1     event_1_arm_1                      2
                    repeat_question_1.repeat_question_1 id.repeat_question_1
1.repeat_question_1                                Good                    1
2.repeat_question_1                                 Bad                    2
                    repeat_question_1.repeat_datetime id.repeat_datetime
1.repeat_question_1               2023-04-20 13:34:00                  1
2.repeat_question_1               2023-04-20 13:35:00                  2
                               name.form
1.repeat_question_1 repeating_instrument
2.repeat_question_1 repeating_instrument
```

**Comment by nutterb** [comment]
In that case, it looks like this is hitting its target. You might be right about the name being confusing. Does this feel like an `exportRecordsByForm`? (I'm spitballing)

**Comment by spgarbet** [comment]
@nutterb @obregos  Where are we on this? Has it passed review?

**Comment by obregos** [comment]
I still need to work on this, it is not giving the expected output. I am expecting the following
```
> widerRepeated(rcon, "record_id")$repeating_instrument
       id.tmp    redcap_event_name   redcap_repeat_instance   repeat_question_1   repeat_datetime                      name.form
1           1     event_1_arm_1             1                                           Good                         2023-04-20 13:34:00 repeating_instrument
2           1     event_1_arm_1             2                                            Bad                          2023-04-20 13:35:00 repeating_instrument  
```

**Comment by obregos** [comment]
`widerRepeated(rcon, "record_id")` should return a list of 5 where the name of each list item is the form name

The output of `widerRepeated(rcon, "record_id")$repeating_instrument` should be a data frame with 2 rows and 6 columns
`widerRepeated(rcon, "record_id")$randomization` should be a data frame with 1 rows and 4 columns
`widerRepeated(rcon, "record_id")$fieldtovar_datetimes` should be a data frame with 2 rows and 15 columns
`widerRepeated(rcon, "record_id")$data_import_export` should be a data frame with 25 rows and 41 columns
`widerRepeated(rcon, "record_id")$branching_logic` should be a data frame with 25 rows and 23 columns

There should be no more than 1 row per record and records with no data should not be included.

For the name change we can keep it as `widerRepeated` or change it to `widerRepeatedForms`

**Comment by spgarbet** [comment]
Suggestion: 

Change `return(list.forms)` at the end to 

`if(length(list.forms) == 1) list.forms[[1]] else list.forms`

Thus any single form sent in comes out as a single form and works with the |> syntax of post processing.

**Comment by spgarbet** [comment]
We need some test cases around this, but it looks right now.

**Comment by spgarbet** [comment]
I accidently merged this before it was finished. (working in wrong branch.).

**Pull request merged**